File: Contest/FindPlayersWithZeroorOneLosses.java
class Solution {
	// TC : O(n)
	// SC : O(Total players O(n)*2)
	public List<List<Integer>> findWinners(int[][] matches) {
		List<List<Integer>> ans = new ArrayList<>();
		Set<Integer> allPlayers = new HashSet<>();

		Map<Integer, Integer> map = new HashMap<>();

		Set<Integer> losers = new HashSet<>();

		Set<Integer> nonLosers = new HashSet<>();

		Set<Integer> lostOneMatch = new HashSet<>();

		for(int[] match: matches){
			allPlayers.add(match[0]);
			allPlayers.add(match[1]);
			losers.add(match[1]);

			int lostTimes = map.getOrDefault(match[1], 0);
			map.put(match[1], lostTimes+1);
		}


		for(Integer player: allPlayers){
			if(!losers.contains(player)){
				nonLosers.add(player);
			}
		}

		for(Map.Entry<Integer, Integer> entry: map.entrySet()) {
			if(entry.getValue() == 1) {
				lostOneMatch.add(entry.getKey());
			}

		}



		List<Integer> nonLosersList = new ArrayList<>(nonLosers);
		Collections.sort(nonLosersList);

		ans.add(new ArrayList<>(nonLosersList));

		List<Integer> lostOneMatchList = new ArrayList<>(lostOneMatch);
		Collections.sort(lostOneMatchList);
		ans.add(lostOneMatchList);


		return ans;

	}
}

File: November2022/C++/find-players-with-zero-or-one-losses.cpp
class Solution {
public:
    vector<vector<int>> findWinners(vector<vector<int>>& matches) {
        
        map<int,int>matchesCount;
        
        int len = matches.size();
    
    vector<vector<int>>ans(2);
        
        for(int index = 0; index < len; index++){
            
            matchesCount[matches[index][0]]++;
            matchesCount[-matches[index][1]]++;
        }
    
    
    for(auto num:matchesCount){
        
        if(num.first<0 && num.second==1){
            ans[1].push_back(-num.first);
        }
        else if(num.first>0  && matchesCount[-num.first]==0){
                 ans[0].push_back(num.first);
        }
    }
    
    sort(ans[0].begin(),ans[0].end());
    sort(ans[1].begin(),ans[1].end());
    
    
    return ans;
    
    }
};